test/unit/master/test_cluster_master.py

```diff
@@ -213,6 +215,29 @@ def test_updating_slave_to_nonexistent_state_should_raise_bad_request_error(self
         with self.assertRaises(BadRequestError):
             master.handle_slave_state_update(slave, 'NONEXISTENT_STATE')
 
+    @genty_dataset (
+        slave_unresponsive=(True,1000,),
+        slave_dead=(False,60,),
+        slave_responsive=(True,60,),
+    )
+    def test_heartbeat_disconnects_unresponsive_slave(self, slave_alive, seconds_since_last_heartbeat):
+        master = ClusterMaster()
+        slave_url = "url"
+        slave = master.connect_slave(slave_url, 1)
+        slave = master.get_slave(int(slave['slave_id']))
+        last_heartbeat_time = datetime.now() - timedelta(seconds=seconds_since_last_heartbeat)
+
+        slave.is_alive = MagicMock(return_value=slave_alive)
+        slave.get_last_heartbeat_time = MagicMock(return_value=last_heartbeat_time)
+        slave.mark_dead = Mock()
+
+        master._disconnect_non_heartbeating_slaves()
```

Comment:
I think there was a comment about this previously, but ideally we don't test via private methods. There are exceptions, but most of the time it's a code smell.